codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customizations/CustomValidationExceptionDecorator.kt

```diff
@@ -0,0 +1,646 @@
+import software.amazon.smithy.rust.codegen.core.smithy.RustSymbolProvider
+import software.amazon.smithy.rust.codegen.core.smithy.protocols.shapeModuleName
+import software.amazon.smithy.rust.codegen.core.util.getTrait
+import software.amazon.smithy.rust.codegen.core.util.targetOrSelf
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
+import software.amazon.smithy.rust.codegen.server.smithy.customize.ServerCodegenDecorator
+import software.amazon.smithy.rust.codegen.server.smithy.generators.BlobLength
+import software.amazon.smithy.rust.codegen.server.smithy.generators.CollectionTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.ConstraintViolation
+import software.amazon.smithy.rust.codegen.server.smithy.generators.Range
+import software.amazon.smithy.rust.codegen.server.smithy.generators.StringTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.UnionConstraintTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.ValidationExceptionConversionGenerator
+import software.amazon.smithy.rust.codegen.server.smithy.generators.isKeyConstrained
+import software.amazon.smithy.rust.codegen.server.smithy.generators.isValueConstrained
+import software.amazon.smithy.rust.codegen.server.smithy.generators.protocol.ServerProtocol
+import software.amazon.smithy.rust.codegen.server.smithy.validationErrorMessage
+import software.amazon.smithy.rust.codegen.server.traits.ValidationExceptionTrait
+import software.amazon.smithy.rust.codegen.server.traits.ValidationFieldListTrait
+import software.amazon.smithy.rust.codegen.server.traits.ValidationFieldMessageTrait
+import software.amazon.smithy.rust.codegen.server.traits.ValidationFieldNameTrait
+import software.amazon.smithy.rust.codegen.server.traits.ValidationMessageTrait
+
+class CustomValidationExceptionDecorator : ServerCodegenDecorator {
+    override val name: String
+        get() = "CustomValidationExceptionDecorator"
+    override val order: Byte
+        get() = 69
+
+    override fun validationExceptionConversion(
+        codegenContext: ServerCodegenContext,
+    ): ValidationExceptionConversionGenerator? {
+        val res = CustomValidationExceptionConversionGenerator(codegenContext)
+        res.customValidationException() ?: return null
+        return res
+    }
+}
+
+class CustomValidationExceptionConversionGenerator(private val codegenContext: ServerCodegenContext) :
+    ValidationExceptionConversionGenerator {
+    companion object {
+        val SHAPE_ID: ShapeId = ShapeId.from("smithy.framework#CustomValidationException")
+    }
+
+    override val shapeId: ShapeId = SHAPE_ID
+
+    fun customValidationException(): StructureShape? {
+        return codegenContext.model.shapes(StructureShape::class.java)
+            .filter { it.hasTrait(ValidationExceptionTrait.ID) }
+            .findFirst()
```

Comment:
what if there are more than one in the model? I'm not positive shape-walking in the model will always happen in a deterministic order. (or it will be deterministic but depends on the import order in the model and other things that shouldn't be load bearing).